watchers/watcher.go

```diff
@@ -0,0 +1,191 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+package watchers
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/types"
+	k8sclientcache "k8s.io/client-go/tools/cache"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	k8scontrollercache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Watcher is an interface that describes the methods provided by a Kubernetes resource watcher.
+type Watcher interface {
+	// Clean cancels the context associated with the watcher, stopping the syncing of resources and freeing up resources.
+	Clean()
+
+	// GetCacheSource returns the syncing source of the watcher's cache.
+	GetCacheSource() (source.SyncingSource, error)
+
+	// GetContextTuple returns the context tuple associated with the watcher.
+	GetContextTuple() (CtxTuple, error)
+
+	// GetName returns the name of the watcher.
+	GetName() string
+
+	// IsExpired returns true if the context associated with the watcher has expired, meaning the watcher is no longer active.
+	IsExpired() bool
+
+	// Start starts the watcher by creating an informer from the watcher's cache and adding a resource event handler to the informer.
+	Start() error
+}
+
+// WatcherConfig holds configuration values used to create a new Watcher instance.
+type WatcherConfig struct {
+	// CacheOptions for configuring the cache.
+	CacheOptions k8scontrollercache.Options
+
+	// Handler function that will be called when an event occurs.
+	Handler k8sclientcache.ResourceEventHandler
+
+	// Log for logging.
+	Log *zap.SugaredLogger
+
+	// Name of the watcher instance.
+	Name string
+
+	// Namespace of the resource to watch.
+	Namespace types.NamespacedName
+
+	// ObjectType of the object to watch.
+	ObjectType client.Object
+}
+
+// CtxTuple is a struct that holds a context and its cancel function, as well as a NamespacedName that identifies a Disruption resource.
+type CtxTuple struct {
+	// CancelFunc is a function that can be called to cancel the associated context.
+	CancelFunc context.CancelFunc
+
+	// Ctx is a context.Context object that is used to manage the lifetime of an operation.
+	Ctx context.Context
```

Comment:
Seconding Matthieu, can't we store the cancelFunc and the namespacedname inside the context, using `context.WithValue` and `ctx.Value` functions??